File: 프로그래머스/2/250136. ［PCCP 기출문제］ 2번 ／ 석유 시추/［PCCP 기출문제］ 2번 ／ 석유 시추.cpp
#include <vector>
#include <algorithm>

using namespace std;

#define MAX_LAND 510
typedef pair<int,int> pp;

bool visited[MAX_LAND][MAX_LAND];
int oils[MAX_LAND];
int N, M, miny, maxy;

// move to left, right, up, down
int dx[4] = {0, 0, -1, 1}, dy[4] = {-1, 1, 0, 0};

int dfs(vector<vector<int>>&land, int x, int y) {
    int totalOilInField = 1;
    
     // 해당 지점 방문 표시
    visited[x][y] = true;

    // 해당 구역이 걸쳐있는 가로줄 구함
    miny = min(y, miny), maxy = max(y, maxy);
    
    for (int i = 0; i < 4; i++) {
        int nx = x + dx[i], ny = y + dy[i];
        if (nx < 0 || N <= nx || ny < 0 || M <= ny || !land[nx][ny] || visited[nx][ny]) continue;
        
        // 현재 구역의 석유 양 계산 
        totalOilInField += dfs(land, nx, ny);
    }

    return totalOilInField;
}

int solution(vector<vector<int>> land) {
    int answer = 0;

    N = land.size(), M = land[0].size();

    for (int i = 0; i < N; i++) {
        for (int j = 0; j < M; j++) {
            if (!land[i][j] || visited[i][j]) continue;
            
            miny = MAX_LAND, maxy = -1;
            
            // 해당 구역의 석유 양 추출 및 해당 구역의 가로 범위 저장
            int amount = dfs(land, i, j);       
            oils[miny] += amount, oils[maxy + 1] -= amount;
        }
    }
    
    int oilsInLine = 0;
    for (int i = 0; i < M; i++) {
        oilsInLine += oils[i];
        answer = max(answer, oilsInLine);
    }
        
    return answer;
}

File: 프로그래머스/2/250136. ［PCCP 기출문제］ 2번 ／ 석유 시추/［PCCP 기출문제］ 2번 ／ 석유 시추.java
import java.util.*;
class Solution {
    int[][] visited;
    int groupNum = 1, n, m, min, max;
    int[] sum;
    public int solution(int[][] land) {
        n = land.length;
        m = land[0].length;
        sum = new int[m + 1];
        visited = new int[n][m];
        
        for (int i = 0; i < n; i++) {
            for (int j = 0; j < m; j++) {
                if (land[i][j] == 1 && visited[i][j] == 0) {
                    min = m;
                    max = 0;
                    visited[i][j] = groupNum;
                    int amount = bfs(land, i, j);
                    
                    groupNum++;
                    sum[min] += amount;
                    sum[max + 1] -= amount;
                }
            }
        }

        int sums = 0;
        int answer = 0;
        for (int i = 0; i < m; i++) {
            sums += sum[i];
            answer = Math.max(sums, answer);
        }
        return answer;
    }
    
    int bfs(int[][] land, int r, int c) {
        Queue<int[]> q = new ArrayDeque<>();
        
        int sum = 0;
        int[][] mv = {{-1, 0}, {1, 0}, {0, -1}, {0, 1}};
        
        q.add(new int[]{r, c});
        while (!q.isEmpty()) {
            sum++;
            int[] now = q.poll();
            r = now[0];
            c = now[1];
            min = Math.min(min, c);
            max = Math.max(max, c);
            
            for (int i = 0; i < 4; i++) {
                int nr = r + mv[i][0];
                int nc = c + mv[i][1];
                if (0 <= nr && nr < n && 0 <= nc && nc < m && visited[nr][nc] == 0 && land[nr][nc] == 1) {
                    visited[nr][nc] = groupNum;
                    q.add(new int[]{nr, nc});
                }
            }    
        }
        
        return sum;
    }
}